File: commons-cli-python/src/main/python/pattern_option_builder.py
import io
import urllib.parse
from main.python.options import Options
from main.python.option import Option
from main.python.java_handler import Types

class PatternOptionBuilder:
    """
    Allows Options to be created from a single String. The pattern contains various single character flags and via an
    optional punctuation character, their expected type.
    """

    # String class
    STRING_VALUE = Types.String

    # Object class
    OBJECT_VALUE = Types.Object

    # Number class
    NUMBER_VALUE = Types.Number

    # Date class
    DATE_VALUE = Types.Date

    # Class class
    CLASS_VALUE = Types.Class

    # File Input Stream class
    EXISTING_FILE_VALUE = Types.FileInputStream

    # File class
    FILE_VALUE = Types.File

    # File array class
    FILES_VALUE = Types.Files

    # URL class
    URL_VALUE = Types.URL

    @staticmethod
    def get_value_class(ch: str):
        """
        Retrieve the class that ch represents.

        :param ch: the specified character
        :return: The class that ch represents
        """
        if ch == '@':
            return PatternOptionBuilder.OBJECT_VALUE
        elif ch == ':':
            return PatternOptionBuilder.STRING_VALUE
        elif ch == '%':
            return PatternOptionBuilder.NUMBER_VALUE
        elif ch == '+':
            return PatternOptionBuilder.CLASS_VALUE
        elif ch == '#':
            return PatternOptionBuilder.DATE_VALUE
        elif ch == '<':
            return PatternOptionBuilder.EXISTING_FILE_VALUE
        elif ch == '>':
            return PatternOptionBuilder.FILE_VALUE
        elif ch == '*':
            return PatternOptionBuilder.FILES_VALUE
        elif ch == '/':
            return PatternOptionBuilder.URL_VALUE
        else:
            return None
    
    @staticmethod
    def is_value_code(ch: str):
        """
        Returns whether ch is a value code, i.e. whether it represents a class in a pattern.

        :param ch: the specified character
        :return: true if ch is a value code, otherwise false.
        """
        return len(ch) == 1 and ch in "@:%+#<>*/!"
    
    @staticmethod
    def parse_pattern(pattern: str):
        """
        Returns the Options instance represented by pattern.
        
        :param pattern: the pattern string
        :return: The Options instance
        """
        opt = ' '
        required = False
        type = None

        options = Options()

        for i in range(len(pattern)):
            ch = pattern[i]

            # a value code comes after an option and specifies
            # details about it
            if not PatternOptionBuilder.is_value_code(ch):
                if opt != ' ':
                    option = Option.builder(str(opt)).has_arg(type != None).required(required).type(type).build()

                    # we have a previous one to deal with
                    options.add_option(option)
                    required = False
                    type = None
                    opt = ' '

                opt = ch
            elif ch == '!':
                required = True
            else:
                type = PatternOptionBuilder.get_value_class(ch)

        if opt != ' ':
            option = Option.builder(str(opt)).has_arg(type != None).required(required).type(type).build()

            # we have a final one to deal with
            options.add_option(option)

        return options

File: commons-cli-graal/src/main/java/org/apache/commons/cli/PatternOptionBuilder.java
/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.cli;

import java.io.File;
import java.io.FileInputStream;
import java.net.URL;
import java.util.Date;

import org.graalvm.polyglot.Value;

/**
 * Allows Options to be created from a single String. The pattern contains various single character flags and via an
 * optional punctuation character, their expected type.
 *
 * <table border="1">
 * <caption>Overview of PatternOptionBuilder patterns</caption>
 * <tr>
 * <td>a</td>
 * <td>-a flag</td>
 * </tr>
 * <tr>
 * <td>b@</td>
 * <td>-b [classname]</td>
 * </tr>
 * <tr>
 * <td>c&gt;</td>
 * <td>-c [filename]</td>
 * </tr>
 * <tr>
 * <td>d+</td>
 * <td>-d [classname] (creates object via empty constructor)</td>
 * </tr>
 * <tr>
 * <td>e%</td>
 * <td>-e [number] (creates Double/Long instance depending on existing of a '.')</td>
 * </tr>
 * <tr>
 * <td>f/</td>
 * <td>-f [url]</td>
 * </tr>
 * <tr>
 * <td>g:</td>
 * <td>-g [string]</td>
 * </tr>
 * </table>
 *
 * <p>
 * For example, the following allows command line flags of '-v -p string-value -f /dir/file'. The exclamation mark
 * precede a mandatory option.
 * </p>
 *
 * <pre>
 * Options options = PatternOptionBuilder.parsePattern("vp:!f/");
 * </pre>
 *
 * <p>
 * TODO These need to break out to OptionType and also to be pluggable.
 * </p>
 */
public class PatternOptionBuilder {


    /**
     * Contains a reference to the Python class.
     */
    private static Value clz = ContextInitializer.getPythonClass("pattern_option_builder.py", "PatternOptionBuilder");

    /** String class */
    public static final Class<String> STRING_VALUE = String.class;

    /** Object class */
    public static final Class<Object> OBJECT_VALUE = Object.class;

    /** Number class */
    public static final Class<Number> NUMBER_VALUE = Number.class;

    /** Date class */
    public static final Class<Date> DATE_VALUE = Date.class;

    /** Class class */
    public static final Class<?> CLASS_VALUE = Class.class;

    /// can we do this one??
    // is meant to check that the file exists, else it errors.
    // ie) it's for reading not writing.

    /** FileInputStream class */
    public static final Class<FileInputStream> EXISTING_FILE_VALUE = FileInputStream.class;

    /** File class */
    public static final Class<File> FILE_VALUE = File.class;

    /** File array class */
    public static final Class<File[]> FILES_VALUE = File[].class;

    /** URL class */
    public static final Class<URL> URL_VALUE = URL.class;

    /**
     * Retrieve the class that {@code ch} represents.
     *
     * @param ch the specified character
     * @return The class that {@code ch} represents
     */
    public static Object getValueClass(final char ch) {
        return clz.invokeMember("get_value_class", ch).as(Object.class);
    }

    /**
     * Returns whether {@code ch} is a value code, i.e. whether it represents a class in a pattern.
     *
     * @param ch the specified character
     * @return true if {@code ch} is a value code, otherwise false.
     */
    public static boolean isValueCode(final char ch) {
        return clz.invokeMember("is_value_code", ch).asBoolean();
    }

    /**
     * Returns the {@link Options} instance represented by {@code pattern}.
     *
     * @param pattern the pattern string
     * @return The {@link Options} instance
     */
    public static Options parsePattern(final String pattern) {
        Value optionsObj = clz.invokeMember("parse_pattern", pattern);
        return new Options(optionsObj);
    }
}
